File: android/app/src/main/kotlin/com/groovelab/groovelab/audio/GrooveLabAudioPlugin.kt
package com.groovelab.groovelab.audio

import android.content.Context
import io.flutter.embedding.engine.plugins.FlutterPlugin
import io.flutter.plugin.common.EventChannel
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel

class GrooveLabAudioPlugin : FlutterPlugin, MethodChannel.MethodCallHandler, EventChannel.StreamHandler {
    private lateinit var methodChannel: MethodChannel
    private lateinit var eventChannel: EventChannel
    private var eventSink: EventChannel.EventSink? = null
    private var metronomeEngine: MetronomeEngine? = null
    private var recordingEngine: RecordingEngine? = null
    private var pedaleraEngine: PedaleraEngine? = null
    private var midiEngine: MIDIEngine? = null
    private var context: Context? = null

    override fun onAttachedToEngine(binding: FlutterPlugin.FlutterPluginBinding) {
        context = binding.applicationContext
        methodChannel = MethodChannel(binding.binaryMessenger, "com.groovelab/audio_engine")
        methodChannel.setMethodCallHandler(this)
        eventChannel = EventChannel(binding.binaryMessenger, "com.groovelab/audio_events")
        eventChannel.setStreamHandler(this)
    }

    override fun onDetachedFromEngine(binding: FlutterPlugin.FlutterPluginBinding) {
        methodChannel.setMethodCallHandler(null)
        eventChannel.setStreamHandler(null)
        metronomeEngine?.stop()
        recordingEngine?.stopRecording()
        pedaleraEngine?.stop()
        midiEngine?.dispose()
    }

    override fun onMethodCall(call: MethodCall, result: MethodChannel.Result) {
        when (call.method) {
            "init" -> {
                metronomeEngine = MetronomeEngine { event ->
                    android.os.Handler(android.os.Looper.getMainLooper()).post {
                        eventSink?.success(event)
                    }
                }
                recordingEngine = RecordingEngine(context!!) { event ->
                    android.os.Handler(android.os.Looper.getMainLooper()).post {
                        eventSink?.success(event)
                    }
                }
                metronomeEngine?.initialize()
                result.success(null)
            }
            "loadSound" -> {
                val key = call.argument<String>("key") ?: return result.error("INVALID_ARGS", "Missing key", null)
                val data = call.argument<ByteArray>("data") ?: return result.error("INVALID_ARGS", "Missing data", null)
                metronomeEngine?.loadSound(key, data)
                result.success(null)
            }
            "playSound" -> {
                val key = call.argument<String>("key") ?: "kick"
                metronomeEngine?.playSingleHit(key)
                result.success(null)
            }
            "startMetronome" -> {
                val bpm = call.argument<Int>("bpm") ?: 120
                val beatsPerBar = call.argument<Int>("beatsPerBar") ?: 4
                val beatUnit = call.argument<Int>("beatUnit") ?: 4
                val subdivision = call.argument<Int>("subdivision") ?: 1
                val swing = call.argument<Int>("swingPercent") ?: 0
                val clickSound = call.argument<String>("clickSound") ?: "Wood"
                val accents = call.argument<List<Double>>("accentPattern") ?: listOf(1.0, 0.7, 0.7, 0.7)
                val haptic = call.argument<Boolean>("hapticEnabled") ?: false
                metronomeEngine?.start(bpm, beatsPerBar, beatUnit, subdivision, swing, clickSound, accents, haptic)
                result.success(null)
            }
            "stopMetronome" -> { metronomeEngine?.stop(); result.success(null) }
            "updateBpm" -> { call.argument<Int>("bpm")?.let { metronomeEngine?.updateBpm(it) }; result.success(null) }
            "updateTimeSignature" -> {
                val beats = call.argument<Int>("beatsPerBar") ?: return result.success(null)
                val unit = call.argument<Int>("beatUnit") ?: return result.success(null)
                metronomeEngine?.updateTimeSignature(beats, unit)
                result.success(null)
            }
            "updateSubdivision" -> { call.argument<Int>("subdivision")?.let { metronomeEngine?.updateSubdivision(it) }; result.success(null) }
            "updateSwing" -> { call.argument<Int>("swingPercent")?.let { metronomeEngine?.updateSwing(it) }; result.success(null) }
            "updateClickSound" -> { call.argument<String>("clickSound")?.let { metronomeEngine?.updateClickSound(it) }; result.success(null) }
            "updateAccentPattern" -> { call.argument<List<Double>>("pattern")?.let { metronomeEngine?.updateAccentPattern(it) }; result.success(null) }
            "setHapticMode" -> { call.argument<Boolean>("enabled")?.let { metronomeEngine?.setHapticMode(it) }; result.success(null) }
            "updateHumanFeel" -> { call.argument<Int>("percent")?.let { metronomeEngine?.updateHumanFeel(it) }; result.success(null) }
            "updatePolyrhythm" -> {
                val enabled = call.argument<Boolean>("enabled") ?: false
                val value = call.argument<Int>("value") ?: 3
                metronomeEngine?.updatePolyrhythm(enabled, value)
                result.success(null)
            }
            "startDrumPattern" -> {
                val bpm = call.argument<Int>("bpm") ?: 120
                val pattern = call.argument<Map<String, List<Int>>>("pattern") ?: emptyMap()
                val swing = call.argument<Int>("swingPercent") ?: 0
                metronomeEngine?.startDrumPattern(bpm, pattern, swing)
                result.success(null)
            }
            "stopDrumPattern" -> { metronomeEngine?.stop(); result.success(null) }
            "updateDrumPattern" -> {
                call.argument<Map<String, List<Int>>>("pattern")?.let { metronomeEngine?.updateDrumPattern(it) }
                result.success(null)
            }
            "updateDrumVolumes" -> {
                call.argument<Map<String, Double>>("volumes")?.let { metronomeEngine?.updateDrumVolumes(it) }
                result.success(null)
            }
            "startRecording" -> { recordingEngine?.startRecording(); result.success(null) }
            "stopRecording" -> { result.success(recordingEngine?.stopRecording()) }
            "enableOnsetDetection" -> {
                val threshold = call.argument<Double>("threshold") ?: 0.1
                val minInterval = call.argument<Int>("minIntervalMs") ?: 50
                recordingEngine?.enableOnsetDetection(threshold, minInterval)
                result.success(null)
            }
            "disableOnsetDetection" -> { recordingEngine?.disableOnsetDetection(); result.success(null) }
            "getOutputLatency" -> { result.success(metronomeEngine?.getOutputLatency() ?: 0.0) }
            "getInputLatency" -> { result.success(recordingEngine?.getInputLatency() ?: 0.0) }
            // ── Pedalera ──
            "initPedalera" -> {
                if (pedaleraEngine == null) {
                    pedaleraEngine = PedaleraEngine(context!!) { event ->
                        android.os.Handler(android.os.Looper.getMainLooper()).post {
                            eventSink?.success(event)
                        }
                    }
                    pedaleraEngine?.initialize()
                }
                result.success(null)
            }
            "setPedalChain" -> {
                @Suppress("UNCHECKED_CAST")
                val chainConfig = call.argument<List<Map<String, Any>>>("chain") ?: emptyList()
                pedaleraEngine?.setChain(chainConfig)
                result.success(null)
            }
            "setPedalParam" -> {
                val idx = call.argument<Int>("index") ?: return result.success(null)
                val name = call.argument<String>("param") ?: return result.success(null)
                val value = call.argument<Double>("value") ?: return result.success(null)
                pedaleraEngine?.setParam(idx, name, value)
                result.success(null)
            }
            "setPedalBypass" -> {
                val idx = call.argument<Int>("index") ?: return result.success(null)
                val bypassed = call.argument<Boolean>("bypassed") ?: return result.success(null)
                pedaleraEngine?.setBypass(idx, bypassed)
                result.success(null)
            }
            "stopPedalera" -> {
                pedaleraEngine?.stop()
                pedaleraEngine = null
                result.success(null)
            }
            "getPedalLatency" -> {
                result.success(pedaleraEngine?.getLatency() ?: 0.0)
            }
            // ── MIDI ──
            "initMidi" -> {
                if (midiEngine == null) {
                    midiEngine = MIDIEngine(context!!) { event ->
                        android.os.Handler(android.os.Looper.getMainLooper()).post {
                            eventSink?.success(event)
                        }
                    }
                }
                result.success(midiEngine?.initialize() ?: false)
            }
            "getMidiDevices" -> { result.success(midiEngine?.listDevices() ?: emptyList<Map<String, Any>>()) }
            "sendMidiNoteOn" -> {
                val note = call.argument<Int>("note") ?: 60
                val velocity = call.argument<Int>("velocity") ?: 100
                val channel = call.argument<Int>("channel") ?: 0
                midiEngine?.sendNoteOn(note, velocity, channel)
                result.success(null)
            }
            "sendMidiNoteOff" -> {
                val note = call.argument<Int>("note") ?: 60
                val channel = call.argument<Int>("channel") ?: 0
                midiEngine?.sendNoteOff(note, channel)
                result.success(null)
            }
            "sendMidiCC" -> {
                val controller = call.argument<Int>("controller") ?: 0
                val value = call.argument<Int>("value") ?: 0
                val channel = call.argument<Int>("channel") ?: 0
                midiEngine?.sendCC(controller, value, channel)
                result.success(null)
            }
            "sendMidiProgramChange" -> {
                val program = call.argument<Int>("program") ?: 0
                val channel = call.argument<Int>("channel") ?: 0
                midiEngine?.sendProgramChange(program, channel)
                result.success(null)
            }
            "startMidiClock" -> {
                val bpm = call.argument<Double>("bpm") ?: 120.0
                midiEngine?.startClock(bpm)
                result.success(null)
            }
            "stopMidiClock" -> { midiEngine?.stopClock(); result.success(null) }
            "updateMidiClockBpm" -> {
                val bpm = call.argument<Double>("bpm") ?: 120.0
                midiEngine?.updateClockBpm(bpm)
                result.success(null)
            }
            "disconnectMidi" -> {
                midiEngine?.dispose()
                midiEngine = null
                result.success(null)
            }
            // ── Audio Session ──
            "getAudioRoute" -> { result.success("speaker") }
            "getAvailableInputs" -> { result.success(emptyList<Map<String, Any>>()) }
            "setPreferredInput" -> { result.success(false) }
            "setBufferDuration" -> { result.success(null) }
            "dispose" -> {
                metronomeEngine?.stop()
                metronomeEngine = null
                recordingEngine?.stopRecording()
                recordingEngine = null
                pedaleraEngine?.stop()
                pedaleraEngine = null
                midiEngine?.dispose()
                midiEngine = null
                result.success(null)
            }
            else -> result.notImplemented()
        }
    }

    override fun onListen(arguments: Any?, events: EventChannel.EventSink?) {
        eventSink = events
    }

    override fun onCancel(arguments: Any?) {
        eventSink = null
    }
}

File: groovelab_backup_20260325/ios/Runner/AudioEngine/GrooveLabAudioPlugin.swift
import Flutter
import UIKit

/// Flutter plugin registration for the native audio engine.
public class GrooveLabAudioPlugin: NSObject, FlutterPlugin {
    private var metronomeEngine: MetronomeEngine?
    private var recordingEngine: RecordingEngine?
    private var pedaleraEngine: PedaleraEngine?
    private var midiEngine: MIDIEngine?
    private var eventSink: FlutterEventSink?

    public static func register(with registrar: FlutterPluginRegistrar) {
        let methodChannel = FlutterMethodChannel(
            name: "com.groovelab/audio_engine",
            binaryMessenger: registrar.messenger()
        )
        let eventChannel = FlutterEventChannel(
            name: "com.groovelab/audio_events",
            binaryMessenger: registrar.messenger()
        )

        let instance = GrooveLabAudioPlugin()
        registrar.addMethodCallDelegate(instance, channel: methodChannel)
        eventChannel.setStreamHandler(instance)
    }

    public func handle(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
        let args = call.arguments as? [String: Any]

        switch call.method {
        case "init":
            initEngine(result: result)
        case "loadSound":
            guard let key = args?["key"] as? String,
                  let data = args?["data"] as? FlutterStandardTypedData else {
                result(FlutterError(code: "INVALID_ARGS", message: "Missing key or data", details: nil))
                return
            }
            metronomeEngine?.loadSound(key: key, wavData: Data(data.data))
            result(nil)
        case "playSound":
            if let key = args?["key"] as? String {
                metronomeEngine?.playSingleHit(key: key)
            }
            result(nil)
        case "startMetronome":
            startMetronome(args: args, result: result)
        case "stopMetronome":
            metronomeEngine?.stop()
            result(nil)
        case "updateBpm":
            if let bpm = args?["bpm"] as? Int {
                metronomeEngine?.updateBpm(bpm)
            }
            result(nil)
        case "updateTimeSignature":
            if let beats = args?["beatsPerBar"] as? Int,
               let unit = args?["beatUnit"] as? Int {
                metronomeEngine?.updateTimeSignature(beatsPerBar: beats, beatUnit: unit)
            }
            result(nil)
        case "updateSubdivision":
            if let sub = args?["subdivision"] as? Int {
                metronomeEngine?.updateSubdivision(sub)
            }
            result(nil)
        case "updateSwing":
            if let pct = args?["swingPercent"] as? Int {
                metronomeEngine?.updateSwing(pct)
            }
            result(nil)
        case "updateClickSound":
            if let sound = args?["clickSound"] as? String {
                metronomeEngine?.updateClickSound(sound)
            }
            result(nil)
        case "updateAccentPattern":
            if let pattern = args?["pattern"] as? [Double] {
                metronomeEngine?.updateAccentPattern(pattern)
            }
            result(nil)
        case "setHapticMode":
            if let enabled = args?["enabled"] as? Bool {
                metronomeEngine?.setHapticMode(enabled)
            }
            result(nil)
        case "updateHumanFeel":
            if let pct = args?["percent"] as? Int {
                metronomeEngine?.updateHumanFeel(pct)
            }
            result(nil)
        case "updatePolyrhythm":
            let enabled = args?["enabled"] as? Bool ?? false
            let value = args?["value"] as? Int ?? 3
            metronomeEngine?.updatePolyrhythm(enabled: enabled, value: value)
            result(nil)
        case "startDrumPattern":
            startDrumPattern(args: args, result: result)
        case "stopDrumPattern":
            metronomeEngine?.stop()
            result(nil)
        case "updateDrumPattern":
            if let pattern = args?["pattern"] as? [String: [Int]] {
                metronomeEngine?.updateDrumPattern(pattern)
            }
            result(nil)
        case "updateDrumVolumes":
            if let volumes = args?["volumes"] as? [String: Double] {
                metronomeEngine?.updateDrumVolumes(volumes)
            }
            result(nil)
        case "startRecording":
            recordingEngine?.startRecording()
            result(nil)
        case "stopRecording":
            let path = recordingEngine?.stopRecording()
            result(path)
        case "enableOnsetDetection":
            let threshold = args?["threshold"] as? Double ?? 0.1
            let minInterval = args?["minIntervalMs"] as? Int ?? 50
            recordingEngine?.enableOnsetDetection(threshold: threshold, minIntervalMs: minInterval)
            result(nil)
        case "disableOnsetDetection":
            recordingEngine?.disableOnsetDetection()
            result(nil)
        case "getOutputLatency":
            let latency = metronomeEngine?.getOutputLatency() ?? 0
            result(latency)
        case "getInputLatency":
            let latency = recordingEngine?.getInputLatency() ?? 0
            result(latency)
        // ── Pedalera ──
        case "initPedalera":
            if pedaleraEngine == nil {
                pedaleraEngine = PedaleraEngine { [weak self] event in
                    DispatchQueue.main.async { self?.eventSink?(event) }
                }
                pedaleraEngine?.initialize()
            }
            result(nil)
        case "setPedalChain":
            if let chainConfig = args?["chain"] as? [[String: Any]] {
                pedaleraEngine?.setChain(chainConfig)
            }
            result(nil)
        case "setPedalParam":
            if let idx = args?["index"] as? Int,
               let name = args?["param"] as? String,
               let val = args?["value"] as? Double {
                pedaleraEngine?.setParam(pedalIndex: idx, paramName: name, value: val)
            }
            result(nil)
        case "setPedalBypass":
            if let idx = args?["index"] as? Int,
               let bypassed = args?["bypassed"] as? Bool {
                pedaleraEngine?.setBypass(pedalIndex: idx, bypassed: bypassed)
            }
            result(nil)
        case "stopPedalera":
            pedaleraEngine?.stop()
            pedaleraEngine = nil
            result(nil)
        case "getPedalLatency":
            result(pedaleraEngine?.getLatency() ?? 0.0)

        // ── CoreMIDI ──
        case "initMidi":
            if midiEngine == nil {
                midiEngine = MIDIEngine { [weak self] event in
                    DispatchQueue.main.async { self?.eventSink?(event) }
                }
            }
            let success = midiEngine?.initialize() ?? false
            result(success)
        case "getMidiDevices":
            result(midiEngine?.listDevices() ?? [])
        case "sendMidiNoteOn":
            if let note = args?["note"] as? Int,
               let velocity = args?["velocity"] as? Int,
               let channel = args?["channel"] as? Int {
                midiEngine?.sendNoteOn(note: UInt8(note), velocity: UInt8(velocity), channel: UInt8(channel))
            }
            result(nil)
        case "sendMidiNoteOff":
            if let note = args?["note"] as? Int,
               let channel = args?["channel"] as? Int {
                midiEngine?.sendNoteOff(note: UInt8(note), channel: UInt8(channel))
            }
            result(nil)
        case "sendMidiCC":
            if let controller = args?["controller"] as? Int,
               let value = args?["value"] as? Int,
               let channel = args?["channel"] as? Int {
                midiEngine?.sendCC(controller: UInt8(controller), value: UInt8(value), channel: UInt8(channel))
            }
            result(nil)
        case "sendMidiProgramChange":
            if let program = args?["program"] as? Int,
               let channel = args?["channel"] as? Int {
                midiEngine?.sendProgramChange(program: UInt8(program), channel: UInt8(channel))
            }
            result(nil)
        case "startMidiClock":
            if let bpm = args?["bpm"] as? Double {
                midiEngine?.startClock(bpm: bpm)
            }
            result(nil)
        case "stopMidiClock":
            midiEngine?.stopClock()
            result(nil)
        case "updateMidiClockBpm":
            if let bpm = args?["bpm"] as? Double {
                midiEngine?.updateClockBpm(bpm)
            }
            result(nil)
        case "disconnectMidi":
            midiEngine?.dispose()
            midiEngine = nil
            result(nil)

        // ── Audio Session / Routing ──
        case "getAudioRoute":
            result(AudioSessionManager.shared.getCurrentRoute())
        case "getAvailableInputs":
            result(AudioSessionManager.shared.getAvailableInputs())
        case "setPreferredInput":
            if let uid = args?["uid"] as? String {
                result(AudioSessionManager.shared.setPreferredInput(uid))
            } else {
                result(false)
            }
        case "setBufferDuration":
            if let ms = args?["durationMs"] as? Double {
                AudioSessionManager.shared.setBufferDuration(ms)
            }
            result(nil)

        case "dispose":
            metronomeEngine?.stop()
            metronomeEngine = nil
            recordingEngine?.stopRecording()
            recordingEngine = nil
            pedaleraEngine?.stop()
            pedaleraEngine = nil
            midiEngine?.dispose()
            midiEngine = nil
            result(nil)
        default:
            result(FlutterMethodNotImplemented)
        }
    }

    private func initEngine(result: FlutterResult) {
        // Configure audio session first
        AudioSessionManager.shared.configure { [weak self] event in
            DispatchQueue.main.async { self?.eventSink?(event) }
        }

        metronomeEngine = MetronomeEngine { [weak self] event in
            DispatchQueue.main.async {
                self?.eventSink?(event)
            }
        }
        recordingEngine = RecordingEngine { [weak self] event in
            DispatchQueue.main.async {
                self?.eventSink?(event)
            }
        }
        metronomeEngine?.initialize()
        result(nil)
    }

    private func startMetronome(args: [String: Any]?, result: FlutterResult) {
        guard let bpm = args?["bpm"] as? Int,
              let beatsPerBar = args?["beatsPerBar"] as? Int,
              let beatUnit = args?["beatUnit"] as? Int else {
            result(FlutterError(code: "INVALID_ARGS", message: "Missing metronome params", details: nil))
            return
        }
        let subdivision = args?["subdivision"] as? Int ?? 1
        let swing = args?["swingPercent"] as? Int ?? 0
        let clickSound = args?["clickSound"] as? String ?? "Wood"
        let accents = args?["accentPattern"] as? [Double] ?? [1.0, 0.7, 0.7, 0.7]
        let haptic = args?["hapticEnabled"] as? Bool ?? false

        metronomeEngine?.start(
            bpm: bpm,
            beatsPerBar: beatsPerBar,
            beatUnit: beatUnit,
            subdivision: subdivision,
            swingPercent: swing,
            clickSound: clickSound,
            accentPattern: accents,
            hapticEnabled: haptic
        )
        result(nil)
    }

    private func startDrumPattern(args: [String: Any]?, result: FlutterResult) {
        guard let bpm = args?["bpm"] as? Int,
              let pattern = args?["pattern"] as? [String: [Int]] else {
            result(FlutterError(code: "INVALID_ARGS", message: "Missing drum params", details: nil))
            return
        }
        let swing = args?["swingPercent"] as? Int ?? 0
        metronomeEngine?.startDrumPattern(bpm: bpm, pattern: pattern, swingPercent: swing)
        result(nil)
    }
}

// MARK: - FlutterStreamHandler
extension GrooveLabAudioPlugin: FlutterStreamHandler {
    public func onListen(withArguments arguments: Any?, eventSink events: @escaping FlutterEventSink) -> FlutterError? {
        self.eventSink = events
        return nil
    }

    public func onCancel(withArguments arguments: Any?) -> FlutterError? {
        self.eventSink = nil
        return nil
    }
}
